session/vpn_creds_sessions_store.go

```diff
@@ -0,0 +1,142 @@
+package session
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/NordSecurity/nordvpn-linux/config"
+	"github.com/NordSecurity/nordvpn-linux/internal"
+)
+
+type VPNCredentialsResponse struct {
+	Username           string
+	Password           string
+	NordLynxPrivateKey string
+}
+
+type VPNCredentialsRenewalAPICall func() (*VPNCredentialsResponse, error)
+
+type vpnCredentialsConfig struct {
+	Username           string
+	Password           string
+	NordLynxPrivateKey string
+}
+
+type VPNCredentialsSessionStore struct {
+	cfgManager         config.Manager
+	errHandlerRegistry *internal.ErrorHandlingRegistry[error]
+	renewAPICall       VPNCredentialsRenewalAPICall
+	externalValidator  VPNCredentialsExternalValidator
+}
+
+// NewVPNCredentialsSessionStore create new VPN credential session store
+func NewVPNCredentialsSessionStore(
+	cfgManager config.Manager,
+	errorHandlingRegistry *internal.ErrorHandlingRegistry[error],
+	renewAPICall VPNCredentialsRenewalAPICall,
+	externalValidator VPNCredentialsExternalValidator,
+) SessionStore {
+	return &VPNCredentialsSessionStore{
+		cfgManager:         cfgManager,
+		errHandlerRegistry: errorHandlingRegistry,
+		renewAPICall:       renewAPICall,
+		externalValidator:  externalValidator,
+	}
+}
+
+// Renew VPN credentials if they have expired.
+func (s *VPNCredentialsSessionStore) Renew() error {
+	if err := s.validate(); err == nil {
+		return nil
+	}
+
+	if s.renewAPICall == nil {
```

Comment:
Ideally, the error should be returned as close as possible to the source of an issue because it's easier to debug. Maybe this could be done at the time of construction?